### ISSUE TITLE
Add 1D-feature mesh build failed

### ISSUE THREAD

**Issue by SYP-code** [CGAL/cgal#6991]
_Please use the following template to help us solving your issue._

## Issue Details

After I apply 1D features to the model and run it for 1 hour, the mesh still cannot be constructed. The mesh construction can be completed in about 3 minutes without adding 1D features. Can you help me see where the problem is in the program? thanks.

## Source Code

#include <vector>
#include <iostream>

#include <CGAL/Mesh_triangulation_3.h>
#include <CGAL/Mesh_complex_3_in_triangulation_3.h>
#include <CGAL/Mesh_criteria_3.h>

#include <CGAL/make_mesh_3.h>
#include <CGAL/Image_3.h>

/// [Domain definition]
#include <CGAL/Exact_predicates_inexact_constructions_kernel.h>
#include <CGAL/Mesh_domain_with_polyline_features_3.h>
#include <CGAL/Labeled_mesh_domain_3.h>

#include "read_polylines.h"
#include <CGAL/Mesh_3/polylines_to_protect.h>  // 未记录的标题

// Domain
typedef CGAL::Exact_predicates_inexact_constructions_kernel K;
typedef CGAL::Labeled_mesh_domain_3<K> Image_domain;
typedef CGAL::Mesh_domain_with_polyline_features_3<Image_domain> Mesh_domain;
/// [Domain definition]

#ifdef CGAL_CONCURRENT_MESH_3
typedef CGAL::Parallel_tag Concurrency_tag;
#else
typedef CGAL::Sequential_tag Concurrency_tag;
#endif

// Triangulation
typedef CGAL::Mesh_triangulation_3<Mesh_domain,CGAL::Default,Concurrency_tag>::type Tr;

typedef CGAL::Mesh_complex_3_in_triangulation_3<Tr> C3t3;

// Criteria
typedef CGAL::Mesh_criteria_3<Tr> Mesh_criteria;

typedef Mesh_criteria::Facet_criteria    Facet_criteria;
typedef Mesh_criteria::Cell_criteria     Cell_criteria;

// To avoid verbose function and named parameters call
using namespace CGAL::parameters;

bool add_1D_features(const CGAL::Image_3& image,
    Mesh_domain& domain)
{
    typedef K::Point_3 Point_3;
    typedef unsigned char Word_type;

    std::vector<std::vector<Point_3> > features_inside;

    std::vector<std::vector<Point_3> > polylines_on_bbox;
    CGAL::polylines_to_protect<Point_3, Word_type>(image, polylines_on_bbox,
        features_inside.begin(),
        features_inside.end());

    domain.add_features(polylines_on_bbox.begin(), polylines_on_bbox.end());

    // It is very important that the polylines from the file `lines_fname`
    // contain only polylines in the inside of the box of the image.
    //domain.add_features(features_inside.begin(), features_inside.end());
    return true;
}
/// [Add 1D features]

int main(int argc, char* argv[])
{
  /// [Loads image]
  //const std::string fname = (argc>1)?argv[1]:CGAL::data_file_path("images/liver.inr.gz");
    const std::string fname = (argc > 1) ? argv[1] : CGAL::data_file_path("images/DTU-ref-256-small.inr");
  CGAL::Image_3 image;
  if(!image.read(fname)){
    std::cerr << "Error: Cannot read file " <<  fname << std::endl;
    return EXIT_FAILURE;
  }
  /// [Loads image]

  /// [Domain creation]
  Mesh_domain domain = Mesh_domain::create_labeled_image_mesh_domain(image);

  if (!add_1D_features(image, domain)) {
      return EXIT_FAILURE;
  }
  // Mesh criteria
 // Mesh_criteria criteria(facet_angle=30, facet_size=6, facet_distance=4,
  //                       cell_radius_edge_ratio=3, cell_size=8);
    Mesh_criteria criteria(edge_size = 10,facet_angle=30, facet_size=10, facet_distance=0.5,
                         cell_radius_edge_ratio=3, cell_size=20);
  //Facet_criteria facet_criteria(30, 10, 0.5); // angle, size, approximation
  //Cell_criteria cell_criteria(2., 20); // radius-edge ratio, size
  //Mesh_criteria criteria(facet_criteria, cell_criteria);
  /// [Meshing]
  C3t3 c3t3 = CGAL::make_mesh_3<C3t3>(domain, criteria);
  /// [Meshing]

  // Output
  std::ofstream medit_file("out.mesh");
  c3t3.output_to_medit(medit_file);

  return 0;
}

##Data
[Data.zip](https://github.com/CGAL/cgal/files/9879694/Data.zip)

## Environment

* Operating system (Windows/64 bits):

**Comment by sloriot** [comment]
What are the labels that you want to mesh and what is the label of the outside?

**Comment by SYP-code** [comment]
> What are the labels that you want to mesh and what is the label of the outside?

Thank you for your reply. My structure is a composite structure made of two or three materials, however the mesh generated with CGAL does not have clear boundaries, I want to add 1D features to get a mesh with sharp boundaries as shown below, but the final mesh cannot be calculated Lattice structure even if the program runs for a long time.

![fig](https://user-images.githubusercontent.com/79560731/198322811-3d338684-c824-44c9-9c63-6b9830955dba.jpg)

**Comment by SYP-code** [comment]
Thank you very much for your guidance, although the small model can be calculated after adding 1D features, but switching to a larger model for 20 minutes still can't calculate the result (it only takes about 3-5 minutes without 1D features), you can help Is there any possibility to optimize the calculation speed?










在 2022-10-27 23:14:35，"Laurent Rineau" ***@***.***> 写道：

Closed #6991 as completed.

**Comment by SYP-code** [comment]
Thank you very much for your guidance, although the small model can be calculated after adding 1D features, but switching to a larger model for 20 minutes still can't calculate the result (it only takes about 3-5 minutes without 1D features), you can help Is there any possibility to optimize the calculation speed?










在 2022-10-27 23:14:35，"Laurent Rineau" ***@***.***> 写道：

Closed #6991 as completed.

**Comment by lrineau** [comment]
How large are the small models, and the larger models? The timings you quote are letting me think you are actually using a Debug build of the code, without any compiler optimization. Can you please confirm you have enabled compiler optimizations in your build?

**Comment by SYP-code** [comment]
I'm not sure if compiler optimizations are turned on, is there a specific guide? How can I do this? Also, I know that there are a lot of functions optimized here in 1D features, how would I add these functions to my existing code? thanks.

**Comment by lrineau** [comment]
> I'm not sure if compiler optimizations are turned on, is there a specific guide? How can I do this? Also, I know that there are a lot of functions optimized here in 1D features, how would I add these functions to my existing code? thanks.

That depends on your build system. In the first message, you mentioned that your operating system is "Windows/64 bits", and so I assume that your are using Visual Studio, with a solution generated from CMake. There is a tool to change the build configuration, in Visual Studio. See that old screenshot:

![CJ0vC](https://user-images.githubusercontent.com/5746675/200005172-1488fa04-b6d6-4d63-8bb2-79b57f6eba0f.png)

(taken from https://stackoverflow.com/q/36623868/1728537)

Make sure you are not using the `Debug` configuration. `Release` or `RelWithDebInfo` will active compiler optimizations.

On the other hand, you have not answered my question:
> How large are the small models, and the larger models?

How large are they? I would like to compare with the timings I have.

**Comment by SYP-code** [comment]
You are right! I did use the Debug configuration instead of the Release build before. Thank you so much Laurent Rineau, this solved a big problem for me. Currently my model is a 181x785x181 matrix with six labels (like the data uploaded below). The computation time without 1D features and with 1D features is 104s and 170s respectively, and the computational efficiency looks great so far! !
However, the mesh generated with 1D features has the problem that the local elements are too refined (as shown in the figure below), which leads to a sharp increase in the number of overall meshes, which will bring a burden to the subsequent finite element calculation. I have followed "Mesh_3 - add detection of intersection lines from labeled images  #6129" which seems to be a good solution to the problem I am having, may I ask how I can improve on my existing code to achieve this powerful functions?

Data: [test.zip](https://github.com/CGAL/cgal/files/9941017/test.zip)

![Mesh_feature](https://user-images.githubusercontent.com/79560731/200047093-e12a17a7-1f09-4617-aaaf-742863b3e811.jpg)

**Comment by janetournois** [comment]
Indeed the PR you mention should solve your problem.
You can use the associated branch, and follow [this example](https://github.com/CGAL/cgal/pull/6129/files?file-filters%5B%5D=.cpp&show-viewed-files=true#diff-7685be21fc3e0ab2fc78f96564ae649d5c09a36bf347ded470e392d57101b878) to use this new feature.

**Comment by SYP-code** [comment]
Thanks for your reply, how would I change the existing header files, Cmake files and other configurations to achieve this function? Can you give further tutorials? Because I am a new user of CGAL and C++, I am relatively new to related operations. Thanks again.
















At 2022-11-07 15:53:07, "Jane Tournois" ***@***.***> wrote:

Indeed the PR you mention should solve your problem.
You can use the associated branch, and follow this example to use this new feature.

**Comment by SYP-code** [comment]
Hello, I replaced the #6129 file with the file of the same name in CGAL5.5, but an error will be reported during the build process of VS. What is the reason? My CGAL is downloaded using the vcpkg package.

**Comment by janetournois** [comment]
There are new headers added in the PR, they should also be added to your copy of CGAL.
The other option is to download the full CGAL branch associated to PR #6129, and use it as CGAL_DIR in CMake
